TrackletAlgorithm/MatchEngineUnit.h

```diff
@@ -101,18 +64,142 @@ class MatchEngineUnit : public MatchEngineUnitBase<VMProjType> {
   stubmask_[3] = nstubsall_[3]!=0;
   ap_uint<2> index = __builtin_ctz(stubmask_);
   stubmask_[index]=0;
-  second_ = isSecond[index];
-  phiPlus_ = isPhiPlus[index];
+  second_ = index[0];
+  phiPlus_ = index[1];
   nstubs_ = nstubsall_[index];
-  ivmphi = projbuffer.getPhi();
-  iphi_ = iphi;
-  auto const qdata=projbuffer_;
-  tcid=qdata.getTCID();
+  iphi_ = projbuffer.getPhi();
+  tcid_ = projbuffer.getTCID();
 
-  good_ = false;
+  good__ = false;
 
 }
 
+
+
+inline void step(const VMStubMECM<VMSMEType> stubmem[2][1024]) {
+#pragma HLS inline
+#pragma HLS array_partition variable=nstubsall_ complete dim=1
+
+  bool nearfull = nearFull();
+
+  good__ = (!idle_) && (!nearfull);
+
+  // Buffer still has projections to read out
+  //If the buffer is not empty we have a projection that we need to 
+  //process. 
+  
+  NSTUBS istubtmp=istub_;
+
+  ap_uint<3> iphiSave = iphi_ + phiPlus_;
+  auto secondSave = second_;
+
+  if(zero<kNBits_MemAddrBinned>()[istub_]) {
+     
+    //Need to read the information about the proj in the buffer
+    VMProjection<BARREL> data(projbuffer_.getProjection());
+
+    //FIXME is this valid? Only using range(3,1) instead of full range, zfirst in MatchProcessor.h
+    zbin = data.getZBin().range(3,1); 
+    
+    auto projfinez = data.getFineZ();
+    projfinephi__ = data.getFinePhi();
+     
+    //Calculate fine z position
+    if (second_) {
+      projfinezadj__ = projfinez-8;
+      zbin=zbin+1;
+    } else {
+      projfinezadj__ = projfinez;
+    }
+
+    if (!phiPlus_) {
+      if (shift_==-1) {
+	projfinephi__ -= 8;
+      }
+    } else {
+      //When we get here shift_ is either 1 or -1
+      if (shift_==1) {
+	projfinephi__ += 8;
+      }
+    }
+
+    isPSseed__ = data.getIsPSSeed();
+    projrinv__ = data.getRInv();
+
+  }
+   
+  //Check if last stub, if so, go to next buffer entry 
+  if (good__) {
+    if (istub_+1>=nstubs_){
+      istub_=0;
+      if (!stubmask_) {
+	idle_ = true;
+      } else {
+	ap_uint<2> index = __builtin_ctz(stubmask_);
+	stubmask_[index]=0;
+	second_ =  index[0];
+	phiPlus_ =  index[1];
+	nstubs_ = nstubsall_[index];
+      }
+    } else {
+      istub_++;
+    }
+  }
+  
+  //Read stub memory and extract data fields
+  ap_uint<10> stubadd=(iphiSave,zbin,istubtmp);
+  stubdata__ = stubmem[bx_&1][stubadd];
+  projbuffer__ = projbuffer_;
+  projseq__ = projseq_;
+
+   
+} // end step
+
+
+
+
+  inline void processPipeLine(ap_uint<1> *table) {
+#pragma HLS inline
+
+    auto stubindex=stubdata___.getIndex();
+    auto stubfinez=stubdata___.getFineZ();
+    auto stubfinephi=stubdata___.getFinePhi();
+    auto stubbend=stubdata___.getBend();
+    
+    bool passphi = isLessThanSize<5,3,false,5,3>()[(projfinephi___,stubfinephi)];
+    
+    //Check if stub z position consistent
+    bool pass = isPSseed___ ? isLessThanSize<5,1,true,3,5>()[(stubfinez,projfinezadj___)] : isLessThanSize<5,5,true,3,5>()[(stubfinez,projfinezadj___)];
```

Comment:
@aryd Can you replace the integer template parameters to `isLessThanSize` with meaningfully named constants?